### PR TITLE
logstash testing

### PR DESCRIPTION
**Pull request by urso**
add more tests + potential fixes to logstash output mode

### CODE REVIEW

**Comment by urso** [comment]
Sorry for big PR. changes in this PR are mostly focused on adding unit tests to output modes and logstash client. All other changes are due to fixes and and refactorings to create unit tests.

- refactor (splitting up) internal functionality into specialised types for adding unit tests:
  - backoff behavior
  - window sizing
  - logstash transport mode (TCP/TLS)
  - logstash protocol vs. client with connection
  - have more logstash client tests running on actual TCP connections instead of fake connection objects (helps with reproducibility)

- fixes
  - logstash client entering infinite loop on too many timeout errors
  - do not retry events that are not json-encodable
  - large timeouts on backoff blocking output plugin shutdown (only for load balancer, single mode still suffers problem)
  - set default window size (bulk_max_size) to 2048 in logstash client (for some reason it was set  back to 1024 in code)
  - raise an error if invalid sequence number in ACK message is received (bigger than max expected)

- other improvements:
  - more detailed debug output in logstash client
  - client unit tests based on test-driver, so unit tests can be re-used with alternative client implementations (e.g. pipelining, new protocol)
  - re-use backoff implementation for balanced and single output mode + add backoff to fail-over mode
  - simplify connection mode initialization based on config
  - simplify load-balancer worker loop
  - load balancer worker forces reconnect/ping if error in send-loop

**Comment by urso** [comment]
for comparison:
- [master codecov](https://codecov.io/github/elastic/beats/libbeat/outputs/logstash?ref=15ff780beed48425e08d5c2921c91863cacc3f37)
- [PR codecov](https://codecov.io/github/elastic/beats/libbeat/outputs/logstash?ref=4e38b54637e2492e11c01eeecb30aa2a9f6da656)

File: libbeat/outputs/logstash/protocol_test.go [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+		tmp, _ := buf.Collect(int(len))
+		buf.Reset()
+
+		dataBuf := streambuf.New(nil)
+		// decompress data
+		decomp, err := zlib.NewReader(streambuf.NewFixed(tmp))
+		if err != nil {
+			return nil, err
+		}
+		// dataBuf.ReadFrom(streambuf.NewFixed(tmp))
```
Comment (andrewkroh) [inline libbeat/outputs/logstash/protocol_test.go]: Do you want to remove this commented out code?

File: libbeat/outputs/logstash/client_test.go [hunk 2]
```diff
@@ -128,8 +86,8 @@ func (m *mockTransport) IsConnected() bool {
 }
 
 func (m *mockTransport) Close() error {
-	close(m.ch)
-	close(m.control)
+	// close(m.ch)
```
Comment (andrewkroh) [inline libbeat/outputs/logstash/client_test.go]: Same here. Do you want to remove this?

**Comment by andrewkroh** [comment]
Looks good!

File: libbeat/outputs/logstash/protocol_test.go (hunk 1 above)
Comment (urso) [inline libbeat/outputs/logstash/protocol_test.go]: Yes, can be removed. thanks

File: libbeat/outputs/logstash/client_test.go (hunk 2 above)
Comment (urso) [inline libbeat/outputs/logstash/client_test.go]: :+1:

**Pull request merged**